File: app/internal/publisher/template/typescript/cache.ts
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

interface CacheEntry {
  value: unknown;
  cached_at: number;
}

interface CacheData {
  [key: string]: CacheEntry;
}

export class CacheManager {
  private config: Record<string, any>;
  private productId: string;
  private _cacheDir: string;
  private _cacheFile: string;
  private _tmpFile: string;
  private _corruptFile: string;
  private _ttlDays: number;
  private _cache: CacheData | null = null;

  constructor(config: Record<string, any>) {
    this.config = config;
    this.productId = config.product?.id || 'unknown';
    const safeName = this.productId.replace(/[^a-zA-Z0-9_\-]/g, '_');
    this._cacheDir = path.join(os.homedir(), '.websmith', safeName);
    this._cacheFile = path.join(this._cacheDir, 'cache.json');
    this._tmpFile = path.join(this._cacheDir, 'cache.tmp');
    this._corruptFile = path.join(this._cacheDir, 'cache.corrupt');
    this._ttlDays = this._getTtl();
  }

  private _getTtl(): number {
    return this.config.offline?.cache_days || 0;
  }

  private _ensureCacheDir(): void {
    fs.mkdirSync(this._cacheDir, { recursive: true });
  }

  private _loadCache(): CacheData {
    if (this._cache !== null) return this._cache;
    this._ensureCacheDir();
    try {
      if (fs.existsSync(this._cacheFile)) {
        const data = fs.readFileSync(this._cacheFile, 'utf-8');
        this._cache = JSON.parse(data) as CacheData;
        return this._cache;
      }
    } catch (_) {
      this._preserveCorruptCache();
    }
    this._cache = {};
    return this._cache;
  }

  private _preserveCorruptCache(): void {
    try {
      if (fs.existsSync(this._cacheFile)) {
        if (fs.existsSync(this._corruptFile)) fs.unlinkSync(this._corruptFile);
        fs.renameSync(this._cacheFile, this._corruptFile);
      }
    } catch (_) {
      try { if (fs.existsSync(this._cacheFile)) fs.unlinkSync(this._cacheFile); } catch (_) {}
    }
  }

  private _saveCache(): void {
    if (this._cache === null) return;
    this._ensureCacheDir();
    try {
      fs.writeFileSync(this._tmpFile, JSON.stringify(this._cache, null, 2), 'utf-8');
      fs.renameSync(this._tmpFile, this._cacheFile);
    } catch (_) {
      try { if (fs.existsSync(this._tmpFile)) fs.unlinkSync(this._tmpFile); } catch (_) {}
    }
  }

  get<T = unknown>(key: string): T | null {
    const cache = this._loadCache();
    const entry = cache[key];
    if (entry === undefined) return null;
    if (this._isExpired(entry)) {
      this.delete(key);
      return null;
    }
    return entry.value as T;
  }

  set(key: string, value: unknown): void {
    const cache = this._loadCache();
    cache[key] = { value, cached_at: Date.now() / 1000 };
    this._saveCache();
  }

  delete(key: string): void {
    const cache = this._loadCache();
    if (key in cache) {
      delete cache[key];
      this._saveCache();
    }
  }

  clear(): void {
    this._cache = {};
    this._saveCache();
  }

  private _isExpired(entry: CacheEntry): boolean {
    const cachedAt = entry.cached_at || 0;
    const ttlSeconds = this._ttlDays * 86400;
    return (Date.now() / 1000 - cachedAt) > ttlSeconds;
  }

  isValid(): boolean {
    const cache = this._loadCache();
    const entry = cache.license_status;
    if (!entry) return false;
    return !this._isExpired(entry);
  }

  exists(): boolean {
    return fs.existsSync(this._cacheFile);
  }

  getLicenseStatus(): Record<string, any> | null {
    return this.get<Record<string, any>>('license_status');
  }

  setLicenseStatus(status: Record<string, any>): void {
    this.set('license_status', status);
  }

  invalidateLicenseStatus(): void {
    this.delete('license_status');
  }

  setOnboardingComplete(): void {
    const cache = this._loadCache();
    cache.onboarding_complete = { value: true, cached_at: Date.now() / 1000 };
    this._saveCache();
  }

  isOnboardingComplete(): boolean {
    return this.get<boolean>('onboarding_complete') === true;
  }

  markHasEverActivatedPaidLicense(): void {
    const cache = this._loadCache();
    cache.has_ever_activated_paid_license = { value: true, cached_at: Date.now() / 1000 };
    this._saveCache();
  }

  hasEverActivatedPaidLicense(): boolean {
    const cache = this._loadCache();
    const entry = cache.has_ever_activated_paid_license;
    return entry !== undefined && entry.value === true;
  }
}

File: app/internal/publisher/template/python/cache.py
"""Local cache manager for license status (offline support)"""
import json
import os
import time
from pathlib import Path
from typing import Any, Dict, Optional


class CacheManager:
    def __init__(self, config: Dict[str, Any]):
        self.config = config
        self.product_id = config.get('product', {}).get('id', 'unknown')
        safe_name = ''.join(c if c.isalnum() or c in '-_' else '_' for c in self.product_id)
        self._cache_dir = Path.home() / '.websmith' / safe_name
        self._cache_file = self._cache_dir / 'cache.json'
        self._tmp_file = self._cache_dir / 'cache.tmp'
        self._corrupt_file = self._cache_dir / 'cache.corrupt'
        self._ttl_days = self._get_ttl()
        self._cache: Optional[Dict[str, Any]] = None

    def _get_ttl(self) -> int:
        offline = self.config.get('offline', {})
        return offline.get('cache_days', 0)

    def _ensure_cache_dir(self) -> None:
        self._cache_dir.mkdir(parents=True, exist_ok=True)

    def _load_cache(self) -> Dict[str, Any]:
        if self._cache is not None:
            return self._cache
        self._ensure_cache_dir()
        if not self._cache_file.exists():
            self._cache = {}
            return self._cache
        try:
            with open(self._cache_file, 'r') as f:
                self._cache = json.load(f)
            return self._cache
        except (json.JSONDecodeError, IOError):
            self._preserve_corrupt_cache()
            self._cache = {}
            return self._cache

    def _preserve_corrupt_cache(self) -> None:
        if self._cache_file.exists():
            try:
                if self._corrupt_file.exists():
                    self._corrupt_file.unlink()
                self._cache_file.rename(self._corrupt_file)
            except Exception:
                self._cache_file.unlink()

    def _save_cache(self) -> None:
        if self._cache is None:
            return
        self._ensure_cache_dir()
        try:
            with open(self._tmp_file, 'w') as f:
                json.dump(self._cache, f, indent=2)
            os.replace(self._tmp_file, self._cache_file)
        except Exception:
            if self._tmp_file.exists():
                try:
                    self._tmp_file.unlink()
                except Exception:
                    pass

    def get(self, key: str) -> Optional[Any]:
        cache = self._load_cache()
        entry = cache.get(key)
        if entry is None:
            return None
        if self.is_expired(entry):
            self.delete(key)
            return None
        return entry.get('value')

    def set(self, key: str, value: Any) -> None:
        cache = self._load_cache()
        cache[key] = {'value': value, 'cached_at': time.time()}
        self._save_cache()

    def delete(self, key: str) -> None:
        cache = self._load_cache()
        if key in cache:
            del cache[key]
            self._save_cache()

    def clear(self) -> None:
        self._cache = {}
        self._save_cache()

    def is_expired(self, entry: Dict[str, Any]) -> bool:
        cached_at = entry.get('cached_at', 0)
        ttl_seconds = self._ttl_days * 24 * 60 * 60
        return (time.time() - cached_at) > ttl_seconds

    def is_valid(self) -> bool:
        cache = self._load_cache()
        entry = cache.get('license_status')
        if entry is None:
            return False
        return not self.is_expired(entry)

    def get_license_status(self) -> Optional[Dict[str, Any]]:
        return self.get('license_status')

    def is_hardware_consistent(self, current_hardware_id: str) -> bool:
        status = self.get_license_status()
        if not status:
            return True
        hardware_id = status.get('hardware_id')
        if not hardware_id:
            return True
        return hardware_id == current_hardware_id

    def invalidate_if_hardware_mismatch(self, current_hardware_id: str) -> None:
        if not self.is_hardware_consistent(current_hardware_id):
            self.invalidate_license_status()

    def set_license_status(self, status: Dict[str, Any]) -> None:
        self.set('license_status', status)

    def invalidate_license_status(self) -> None:
        self.delete('license_status')

    def save_license_key(self, license_key: str) -> None:
        key_path = self._cache_dir / 'license.key'
        try:
            key_path.write_text(license_key.strip())
        except Exception:
            pass

    def load_license_key(self) -> Optional[str]:
        key_path = self._cache_dir / 'license.key'
        if key_path.exists():
            try:
                return key_path.read_text().strip() or None
            except Exception:
                pass
        return None

    def clear_license_key(self) -> None:
        key_path = self._cache_dir / 'license.key'
        if key_path.exists():
            try:
                key_path.unlink()
            except Exception:
                pass

    def set_onboarding_complete(self) -> None:
        cache = self._load_cache()
        cache['onboarding_complete'] = {'value': True, 'cached_at': time.time()}
        self._save_cache()

    def is_onboarding_complete(self) -> bool:
        return self.get('onboarding_complete') is True

    def mark_has_ever_activated_paid_license(self) -> None:
        cache = self._load_cache()
        cache['has_ever_activated_paid_license'] = {'value': True, 'cached_at': time.time()}
        self._save_cache()

    def has_ever_activated_paid_license(self) -> bool:
        return self.get('has_ever_activated_paid_license') is True

    # ====================================================================
    # Message Queue (Offline Retry)
    # ====================================================================

    def queue_message(self, msg: Dict[str, Any]) -> None:
        queue = self.get_message_queue()
        msg['id'] = msg.get('id', f"q_{int(time.time())}_{os.urandom(4).hex()}")
        msg['status'] = msg.get('status', 'pending')
        msg['retry_count'] = msg.get('retry_count', 0)
        msg['max_retries'] = msg.get('max_retries', 5)
        msg['created_at'] = msg.get('created_at', int(time.time()))
        msg['next_retry_at'] = msg.get('next_retry_at', int(time.time()) + 60)
        queue.append(msg)
        self.set('message_queue', queue)

    def get_message_queue(self) -> list:
        return self.get('message_queue') or []

    def save_message_queue(self, queue: list) -> None:
        self.set('message_queue', queue)

    def cleanup_sent_messages(self) -> None:
        queue = [m for m in self.get_message_queue() if m.get('status') != 'sent']
        self.save_message_queue(queue)

    def get_pending_count(self) -> int:
        return len([m for m in self.get_message_queue() if m.get('status') in ('pending', 'failed')])

    def reset_all(self) -> None:
        self.clear()
        self.clear_license_key()
